src/control/pure_pursuit_cpp/src/node_pure_pursuit_cpp.cpp

```diff
@@ -0,0 +1,164 @@
+#include <tf2/LinearMath/Matrix3x3.h>
+#include <tf2/LinearMath/Quaternion.h>
+
+#include <chrono>
+
+#include "ackermann_msgs/msg/ackermann_drive_stamped.hpp"
+#include "driverless_msgs/msg/path_stamped.hpp"
+#include "geometry_msgs/msg/pose_with_covariance_stamped.hpp"
+#include "rclcpp/rclcpp.hpp"
+
+using std::placeholders::_1;
+
+struct Pose {
+    double const x;
+    double const y;
+    double const yaw;
+
+    Pose(double x, double y, double yaw) : x(x), y(y), yaw(yaw) {}
+};
+
+struct Point {
+    double const x;
+    double const y;
+    double const turn_intensity;
+
+    Point(double x, double y, double turn_intensity) : x(x), y(y), turn_intensity(turn_intensity) {}
+};
+
+double angle(double x1, double y1, double x2, double y2) {
+    auto x_disp = x2 - x1;
+    auto y_disp = y2 - y1;
+    return atan2(y_disp, x_disp);
+}
+
+// https://stackoverflow.com/a/29871193
+double wrap_to_pi(double x) {
+    auto min = x - (-M_PI);
+    auto max = M_PI - (-M_PI);
+    return -M_PI + fmod(max + fmod(min, max), max);
+}
+
+class Fast : public rclcpp::Node {
+   private:
+    rclcpp::Subscription<driverless_msgs::msg::PathStamped>::SharedPtr path_sub;
+    rclcpp::Subscription<geometry_msgs::msg::PoseWithCovarianceStamped>::SharedPtr slam_pose_sub;
+    rclcpp::Publisher<ackermann_msgs::msg::AckermannDriveStamped>::SharedPtr driving_command_pub;
+    std::vector<Point> path{};
+    double lookahead;
+    double squared_lookahead;  // allows us to skip performing sqrts in get_rvwp()
+    double kp_ang;
+    double target_velocity;
+    size_t count{0};
+
+    static Pose calc_car_axle_pos(double car_pos_x, double car_pos_y, double heading) {
+        constexpr double cog2axle = 0.5;
+        double x = car_pos_x + cos(heading) * cog2axle;
+        double y = car_pos_y + sin(heading) * cog2axle;
+        return {x, y, heading};
+    }
+
+    Point get_rvwp(Pose const& car_axle_pos) const {
+        auto closest_dist = DBL_MAX;
+        int closest_index = -1;
+        for (size_t i = 0; i < path.size(); i++) {
+            auto const& point = path.at(i);
+            double diff_x = point.x - car_axle_pos.x;
+            double diff_y = point.y - car_axle_pos.y;
+            double dist = diff_x * diff_x + diff_y * diff_y;
+            if (dist < closest_dist) {
+                closest_dist = dist;
+                closest_index = i;
+            }
+        }
+        assert(closest_index != -1);
```

Comment:
Realistically, this would never be possible as path.size() will always be greater than 0 (see line 108 for checking that)